autode/opt/optimisers/robust.py

```diff
@@ -0,0 +1,458 @@
+                                    (from ``NDOptimiser``)
+
+        See Also:
+            :py:meth:`NDOptimiser <NDOptimiser.__init__>`
+        """
+        super().__init__(*args, init_alpha=init_trust, **kwargs)
+
+        assert 0.0 < min_trust < max_trust
+        self._max_alpha = float(max_trust)
+        self._min_alpha = float(min_trust)
+        self._upd_alpha = bool(update_trust)
+
+        if coord_type.lower() in ["cart", "cartesian"]:
+            self._coord_type = "cart"
+        elif coord_type.lower() == "dic":
+            self._coord_type = "dic"
+        else:
+            raise ValueError("Coordinate type must be either 'cart' or 'DIC'")
+        # todo stop if any constraints detected -> is this needed
+        self._should_damp = bool(damp)
+        self._last_damp_iter = 0
+
+        self._hessian_update_types = [FlowchartUpdate]
+
+    @property
+    def _g_norm(self) -> GradientRMS:
+        """Calculate the RMS gradient norm in Cartesian coordinates"""
+        grad = self._coords.to("cart").g
+        return GradientRMS(np.sqrt(np.average(np.square(grad))))
+
+    @property
+    def converged(self) -> bool:
+        if self._species is not None and self._species.n_atoms == 1:
+            return True  # Optimisation 0 DOF is always converged
+
+        # gradient is better indicator of stationary point
+        if self._g_norm < self.gtol / 10:
+            logger.warning(
+                f"Gradient norm criteria overachieved. "
+                f"{self._gtol.to('Ha/ang')/10:.3f} Ha/Å. "
+                f"Signalling convergence"
+            )
+
+        # both gradient and energy must be converged!!
+        return self._abs_delta_e < self.etol and self._g_norm < self.gtol
+
+    def _initialise_run(self) -> None:
+        self._build_coordinates()
+        self._coords.update_h_from_cart_h(self._low_level_cart_hessian)
+        self._update_gradient_and_energy()
+
+    def _build_coordinates(self) -> None:
+        """Build delocalised internal coordinates"""
+        cart_coords = CartesianCoordinates(self._species.coordinates)
+        if self._coord_type == "cart":
+            self._coords = cart_coords
+
+        else:
+            primitives = self._primitives
+            if len(primitives) < cart_coords.expected_number_of_dof:
+                logger.info(
+                    "Had an incomplete set of primitives. Adding "
+                    "additional distances"
+                )
+                for i, j in combinations(range(self._species.n_atoms), 2):
+                    primitives.append(Distance(i, j))
+            self._coords = DIC.from_cartesian(
+                x=cart_coords, primitives=primitives
+            )
+
+        return None
+
+    def _step(self) -> None:
+
+        self._coords.h = self._updated_h()
+        self._update_trust_radius()
+
+        # damping sets coords so return early
+        if self._damp_if_required():
```

Comment:
inline comments are bad 😉 
```suggestion
        if self._should_damp:
            self._damped_step()
            return
```

Reply:
But I need to know if the damping took place or not, that's why I have the damping function return True or False. The should_damp flag is just the user's choice, but the energy/gradient oscillation also needs to be checked.

Reply:
ah okay, `should_damp` reads like it's a calculated thing rather than a config option.. 
thinking about it, why would you not want to allow damped steps?
why don't you want to separate the oscillation check and taking a damped step?

Reply:
Separated oscillation check and damped step